### PR TITLE
fix(fileio): replace characters over INT_MAX with U+FFFD

### LINKED ISSUE #11877: Assertion failure editing invalid UTF-32
- `nvim --version`: NVIM v0.5.0-23-gafd576ee9
- `vim -u DEFAULTS` (version: ) behaves differently? yes
- Operating system/version: Ubuntu 18.0.4.1 via WSL
- Terminal name/version: conhost
- `$TERM`: xterm-256color

### Steps to reproduce using `nvim -u NORC`

Download: 
[neovim_enc_bug.txt](https://github.com/neovim/neovim/files/4208811/neovim_enc_bug.txt)

```
nvim -u NORC neovim_enc_bug.txt
:e ++enc=utf-32
```

### Actual behaviour
```
nvim: ../src/nvim/fileio.c:1427: int readfile(char_u *, char_u *, linenr_T, linenr_T, linenr_T, exarg_T *, int): Assertion `u8c <= INT_MAX' failed.
```

### Expected behaviour
No crash. (Vim 8 does not crash.)

### PR DESCRIPTION
**Pull request by muniter**
Just pushing forward the previous PR adding the tests.

fixes #11877
credit: @zubairabid https://github.com/neovim/neovim/pull/12010

### CODE REVIEW

File: test/functional/core/fileio_spec.lua [hunk 1]
```diff
@@ -128,5 +129,12 @@ describe('fileio', function()
     table.insert(text, '')
     eq(text, funcs.readfile(fname, 'b'))
   end)
+  it('read invalid u8 over INT_MAX doesn\'t segfault', function()
+    clear()
+    command('call writefile(0zFFFFFFFF, "Xtest-u8-int-max")')
+    local fname = 'Xtest-u8-int-max'
+    -- This should not segfault
+    command("edit ++enc=utf32 "..fname)
```
Comment (gpanders) [inline test/functional/core/fileio_spec.lua:137]: You can use the `assert_alive()` helper (need to import it at the top level):

```suggestion
    command("edit ++enc=utf32 "..fname)
    assert_alive()
```

Comment (gpanders) [inline test/functional/core/fileio_spec.lua:137]: You can use the `assert_alive()` helper (need to import it at the top level):

```suggestion
    -- This should not segfault
    command("edit ++enc=utf32 "..fname)
    assert_alive()
```

**Comment by muniter** [comment]
@gpanders Have updated with the recommendations, CI failures are unrelated.

macos-10.15 timeout
```
macos-10.15  (cc=clang)	Build and test	2021-11-24T02:14:06.1308730Z -- Tests exited non-zero: Process terminated due to timeout
```

macos-11.0 put spec.
```
macos-11.0  (cc=clang)	Build and test	2021-11-24T01:48:21.3294670Z [ RUN      ] put command blockwise register double gP pastes before cursor and leaves cursor after text: 2.66 ms ERR
macos-11.0  (cc=clang)	Build and test	2021-11-24T01:48:21.3295540Z test/helpers.lua:73: Expected objects to be the same.
macos-11.0  (cc=clang)	Build and test	2021-11-24T01:48:21.3296040Z Passed in:
macos-11.0  (cc=clang)	Build and test	2021-11-24T01:48:21.3297340Z (string) 'test_stringbtest_stringbLine of words 1
macos-11.0  (cc=clang)	Build and test	2021-11-24T01:48:21.3297930Z test_stringbtest_stringbLine of words 2
macos-11.0  (cc=clang)	Build and test	2021-11-24T01:48:21.3298800Z test_stringbtest_stringb'
macos-11.0  (cc=clang)	Build and test	2021-11-24T01:48:21.3299220Z Expected:
macos-11.0  (cc=clang)	Build and test	2021-11-24T01:48:21.3300070Z (string) 'test_stringbtest_stringbLine of words 1
macos-11.0  (cc=clang)	Build and test	2021-11-24T01:48:21.3300650Z test_stringbtest_stringbLine of words 1
macos-11.0  (cc=clang)	Build and test	2021-11-24T01:48:21.3301200Z test_stringbtest_stringbLine of words 2
macos-11.0  (cc=clang)	Build and test	2021-11-24T01:48:21.3302060Z test_stringbtest_stringb'
macos-11.0  (cc=clang)	Build and test	2021-11-24T01:48:21.3303040Z ------------------------------------------------------------------------------
```

**Comment by gpanders** [comment]
Going to assume @bfredl's [opinion on the code changes](https://github.com/neovim/neovim/pull/12010#issuecomment-600223262) hasn't changed. Thanks for writing the test.

**Pull request merged**